keps/sig-scheduling/5007-device-attach-before-pod-scheduled/README.md

```diff
@@ -0,0 +1,1114 @@
+These fields will contain a list of conditions, each representing a specific state or event related to the device.
+
+For this feature, following fields are added:
+
+```go
+// AllocatedDeviceStatus contains the status of an allocated device, if the
+// driver chooses to report it. This may include driver-specific information.
+type AllocatedDeviceStatus struct {
+    ...
+    // BindingGates defines the gates for binding.
+    //
+    // +optional
+    BindingGates map[string]bool
+
+    // BindingFailureGates defines the gates for binding failure.
+    //
+    // +optional
+    BindingFailureGates map[string]bool
+}
+```
+
+#### Scheduler DRA Plugin Modifications
+
+When `UsageRestrictedToNode: true` is set, the scheduler DRA plugin will perform the following steps:
+
+1. **Set NodeSelector**: Before the `PreBind` phase, add the `NodeName` to the `ResourceClaim`'s `NodeSelector`.
+
+If Gates are present, the scheduler DRA plugin will perform the following steps during the `PreBind` phase:
+
+2. **Copy Gates**: Copy `BindingGates` and `BindingFailureGates` from `ResourceSlice.Device.Basic` to `AllocatedDeviceStatus`.
+3. **Wait for Conditions**: Wait for the following conditions:
+   - If `NeedToPreparing` is `True`, wait until `IsPrepared` is `True` before proceeding to Bind.
+   - If `PreparingFailed` is `True`, clear the allocation in the `ResourceClaim` and reschedule the Pod.
+   - If the preparation takes longer than the `PreparingTimeout` period, clear the allocation in the `ResourceClaim` and reschedule the Pod.
+
+To support these steps, the following keys are defined:
+
+```go
+const (
+    // NeedToPreparing indicates that this device needs some preparation.
+    // If this flag is True, the scheduler waits in PreBind.
+    NeedToPreparing = "kubernetes.io/need-to-preparing"
+
+    // IsPrepared indicates the device ready state.
+    // If NeedToPreparing is True and IsPrepared is True, the scheduler proceeds to Bind.
+    IsPrepared = "kubernetes.io/is-prepared"
+
+    // PreparingFailed indicates the device preparation failed state.
+    // If PreparingFailed is True, the scheduler will clear the allocation in the ResourceClaim and reschedule the Pod.
+    PreparingFailed = "kubernetes.io/preparing-failed"
+
+    // PreparingTimeout indicates the prepare timeout period.
+    // If the timeout period is exceeded, the scheduler clears the allocation in the ResourceClaim and reschedules the Pod.
+    PreparingTimeout = "kubernetes.io/preparing-timeout"
+)
+```
+
+Note: There is a concern that the in-flight events cache may grow too large when waiting in PreBind.
+To address this, the scheduling framework will be modified to flush the in-flight events cache before PreBind.
+This prevents issues in the scheduling queue caused by keeping pods at PreBind for an extended period.
+This issue will be addressed separately as outlined in kubernetes/kubernetes#129967.
+
+#### PreBind Phase Timeout
```

Comment:
We should decide if we need a hard timeout, or can just use whatever value is given to us. We can also consider having a default timeout if it's not present in the ResourceSlice.

A hard timeout puts a limit on the utility. I would probably prefer having a default, but allowing the vendor to specify an arbitrarily long timeout in the ResourceSlice. So long as that is not disruptive to other pods trying to schedule.

In other words, 5m could be a default, but not hard coded.